core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,289 @@
+package filetransfer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/wandb/wandb/core/internal/observability"
+	"golang.org/x/sync/errgroup"
+)
+
+var maxS3Workers int = 500
+var s3Scheme string = "s3"
+
+// S3FileTransfer uploads or downloads files to/from s3
+type S3FileTransfer struct {
+	// client is the HTTP client for the file transfer
+	client *s3.Client
+
+	// logger is the logger for the file transfer
+	logger *observability.CoreLogger
+
+	// fileTransferStats is used to track upload/download progress
+	fileTransferStats FileTransferStats
+
+	ctx context.Context
+}
+
+// News3FileTransfer creates a new fileTransfer
+func NewS3FileTransfer(
+	client *s3.Client,
+	logger *observability.CoreLogger,
+	fileTransferStats FileTransferStats,
+) (*S3FileTransfer, error) {
+	ctx := context.TODO()
```

Comment:
NOTE: TODO() is more for development and we shouldn't generally merge PRs with it. Background() is appropriate for "root-level" contexts, but see my other comment.